### DISCUSSION TITLE
Outsourcing modules? Integration with other plugins

### DISCUSSION THREAD

**Discussion by HyunggyuJang** [agzam/spacehammer#134] [Ideas]
Hi! I overjoyed by finding this superb project! I love the philosophy with which this set of package made, and I&rsquo;m very happy, as an *Emacs* and Nix user, it will complete my missing part of current environment, that is, the managing the overall dynamic behavior outside of Emacs. I think the *static* one is handled by Nix for good.

Currently I&rsquo;m using the combination of yabai + skhd, however, which mostly covers only for the window and some quick jumps to the specific applications, not the rest. So for the rest, e.g. key remapping, etc. should be handled yet another application, in my case, karabiner.

To integrate those decentralized ones into one ecosystem, I thought Hammerspoon will do the right job for me, meanwhile, I also found this repository. I love the idea exploiting Lisp&rsquo;s flexibility and willing to learn fennel for this amazing ecosystem, spacehammer, so I&rsquo;m thinking about to migrate to this over other combination of applications, which lack of consistency.

My question is, as a newbie for the Lua and Hammerspoon, is it feasible or planned to incorporate other *Spoons*, say [hhtwm](https://github.com/szymonkaliski/hhtwm)? What I&rsquo;m missing most is the yabai&rsquo;s tiling feature and the ability to move an application to a specified space, which seems not implemented in spacehammer (please teach me if I&rsquo;m wrong about it!).

Plus, I believe allowing other existing plugins to be merged into spacehammer easily, it can be a huge momentum for it.

Thanks for your awesome contribution!

Best regards,

&#x2013; **Hyunggyu Jang**

**Comment by agzam** [comment]
Thank you for your kind words, but I have to be honest here - it is far from being qualified to be called "superb" :).

Frankly, this thing, in the beginning, was just a tiny Hammerspoon config, and I kept rolling more and more stuff on top of it. It was never intended to be used by anyone else. I've never thought it would turn into a "real project". It still suffers from some decisions I made early on.

At one point, @eccentric-j  joined and revamped the entire state-transition system, brought many improvements and fixes, opened new possibilities.

Unfortunately, it has become challenging for me to find enough time to work on it in the past months and years. I have many ideas circulating in my head, and I'm hoping to get some time to do some serious hacking, add more features, make it more customizable, modular, etc.; Add REPL, add tests, improve performance, and more.

I am feared for that "huge momentum" you're talking about to a certain degree because I think the project in its current form is nowhere near to be ready for mainstream use. However, I promise you - I won't abandon it unless I build something better.

**Comment by HyunggyuJang** [comment]
You&rsquo;re being humble; it is indeed great as it is! I love its philosophy, and feel home as I&rsquo;m using doom, which is also built on similar creed.

Seems like I&rsquo;d better to start with vanilla hammerspoon packages alongside yours, until I get the idea how to build my own configuration on spacehammer. I&rsquo;m looking forward to the features and ideas you mentioned, especially, among those, REPL. Think it is crucial as it brought big win to Emacs also: one can easily experiments with one&rsquo;s crude ideas and can ask the system what&rsquo;s going on afterward.

The last word relieves me. I&rsquo;ll keep my eyes on this, and if I have such time and energy, I want to make your ideas come to be real.

**Comment by jaidetree** [comment]
@HyunggyuJang We just merged a fix that fixes compatibility with spoons #123 so you could add some spacehammer customizations to use `hhtwm`. 

In #119 we are discussing some ideas for tiling window manager. I've got one idea in mind where we can create composable layouts with specific apps. It will be a big challenge, but should play nicely with OS X. 

Will need to play with `hhtwm` though to see how that feels

**Comment by HyunggyuJang** [comment]
What a huge improvement! Thanks for your elaboration @eccentric-j!  

At the moment, using the combination of yabai and skhd, but thinking to migrate to Spacehammer as of now, the REPL feature implemented!  

I&rsquo;ve skimmed recent achievements and realized novel features are done, like `advice` feature, and so on.  

Thanks for heads up. I&rsquo;ll give a shot on my free time!
